# 01 — Exploratory Data Analysis

### Project: Online Retail Engagement Risk Prediction
### Notebook: 01_eda.ipynb

**Goals:**
- Understand the raw transaction data structure
- Assess data quality and perform cleaning
- Construct an order-level event table
- Establish dataset time range and basic statistics

**Inputs:**
- `online_retail_II.xlsx` — UCI Online Retail Dataset

**Outputs:**
- `data/processed/orders.parquet` — cleaned order-level table

**Steps:**
- Load and inspect raw transaction data
- Clean data — remove missing IDs, returns, and cancellations
- Construct order-level aggregation table
- Persist orders table to parquet
- Analyze revenue and order frequency distributions

In [10]:
# ─────────────────────────────────────────────────────────────
# Load Raw Data
# ─────────────────────────────────────────────────────────────

import pandas as pd
from pathlib import Path

DATA_PATH = Path("../data/raw")
files = list(DATA_PATH.glob("*.xls*"))

df = pd.read_excel(files[0])

print("Shape:", df.shape)
print("Columns:", df.columns.tolist())
print("Date range:", df["InvoiceDate"].min(), "→", df["InvoiceDate"].max())
print("Unique customers:", df["Customer ID"].nunique())

df.head()

Shape: (525461, 8)
Columns: ['Invoice', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'Price', 'Customer ID', 'Country']
Date range: 2009-12-01 07:45:00 → 2010-12-09 20:01:00
Unique customers: 4383


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


## Dataset Description

The raw dataset contains **525,461 transactions** across **8 columns**, spanning
early December 2009 to early December 2010, with **4,383 unique customers**.

We define a customer engagement event as: **one invoice = one order event.**

Each row represents a single line item within an order. Orders are identified
by a unique Invoice ID and will be aggregated to the order level in the next step.

In [11]:
# ─────────────────────────────────────────────────────────────
# Data Cleaning
# - Drop rows with missing Customer ID
# - Keep only positive purchases (exclude returns/cancellations)
# - Compute revenue per line item
# - Normalize data types
# ─────────────────────────────────────────────────────────────

df_clean = df.copy()

# Drop rows without a Customer ID
df_clean = df_clean.dropna(subset=["Customer ID"])

# Keep only positive purchases (exclude returns/cancellations)
df_clean = df_clean[df_clean["Quantity"] > 0]
df_clean = df_clean[df_clean["Price"] > 0]

# Revenue per line item
df_clean["Revenue"] = df_clean["Quantity"] * df_clean["Price"]

# Normalize data types
df_clean["Customer ID"] = df_clean["Customer ID"].astype(int).astype(str)
df_clean["Invoice"]     = df_clean["Invoice"].astype(str)
df_clean["InvoiceDate"] = pd.to_datetime(df_clean["InvoiceDate"])

print("Clean dataset shape:", df_clean.shape)
df_clean.head()

Clean dataset shape: (407664, 9)


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country,Revenue
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085,United Kingdom,83.4
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085,United Kingdom,81.0
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085,United Kingdom,81.0
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085,United Kingdom,100.8
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085,United Kingdom,30.0


After cleaning, the dataset contains **407,664 transactions** across 9 columns.
Removed rows account for missing customer IDs, returns (negative quantity),
and zero-price records.

## Diagnosing Filtered Records

In [13]:
# ─────────────────────────────────────────────────────────────
# Quantify Excluded Records
# Break down the ~118k filtered rows by exclusion reason
# to assess whether we're discarding useful signal.
# ─────────────────────────────────────────────────────────────

# Reload raw data for comparison (df is already in memory from earlier)
null_id    = df["Customer ID"].isna()
cancel     = df["Invoice"].astype(str).str.startswith("C")
non_pos_qty = df["Quantity"] <= 0
zero_price = df["Price"] <= 0

print(f"Total raw rows:              {len(df):,}")
print(f"Null Customer ID:            {null_id.sum():,}")
print(f"Cancellations (C-prefix):    {cancel.sum():,}")
print(f"Non-positive quantity:       {non_pos_qty.sum():,}")
print(f"Zero/negative price:         {zero_price.sum():,}")
print(f"\n--- Overlap ---")
print(f"Null ID + cancellation:      {(null_id & cancel).sum():,}")
print(f"Null ID + non-pos qty:       {(null_id & non_pos_qty).sum():,}")
print(f"Cancel + non-pos qty:        {(cancel & non_pos_qty).sum():,}")
print(f"All three:                   {(null_id & cancel & non_pos_qty).sum():,}")

Total raw rows:              525,461
Null Customer ID:            107,927
Cancellations (C-prefix):    10,206
Non-positive quantity:       12,326
Zero/negative price:         3,690

--- Overlap ---
Null ID + cancellation:      367
Null ID + non-pos qty:       2,487
Cancel + non-pos qty:        10,205
All three:                   366


In [14]:
print(df[non_pos_qty & ~cancel & ~null_id].shape)
print(df[non_pos_qty & ~cancel]["Customer ID"].isna().value_counts())

(0, 8)
Customer ID
True    2121
Name: count, dtype: int64


In [15]:
# ─────────────────────────────────────────────────────────────
# Customer Overlap: Who is cancelling?
# ─────────────────────────────────────────────────────────────

# Identifiable cancellations (have a Customer ID + C-prefix)
id_cancel = df[cancel & ~null_id]

# Current gross purchase customers
gross_customers = set(df_clean["Customer ID"].unique())
cancel_customers = set(id_cancel["Customer ID"].astype(int).astype(str).unique())

print(f"Gross purchase customers:      {len(gross_customers):,}")
print(f"Customers with cancellations:  {len(cancel_customers):,}")
print(f"\n--- Overlap with gross customers ---")
print(f"Cancel customers in gross set: {len(cancel_customers & gross_customers):,} "
      f"({len(cancel_customers & gross_customers)/len(cancel_customers):.1%})")

Gross purchase customers:      4,312
Customers with cancellations:  1,798

--- Overlap with gross customers ---
Cancel customers in gross set: 1,729 (96.2%)


In [17]:
# ─────────────────────────────────────────────────────────────
# Sanity Check — Date Range
# ─────────────────────────────────────────────────────────────

print("Earliest order:", orders["order_ts"].min())
print("Latest order:  ", orders["order_ts"].max())

Earliest order: 2009-12-01 07:45:00
Latest order:   2010-12-09 20:01:00


In [18]:
# ─────────────────────────────────────────────────────────────
# Persist Orders Table
# Saved to data/processed/ — excluded from version control
# via .gitignore. Run this notebook to regenerate.
# ─────────────────────────────────────────────────────────────

out = Path("../data/processed")
out.mkdir(parents=True, exist_ok=True)

orders.to_parquet(out / "orders.parquet", index=False)
print("Saved to:", out / "orders.parquet")
print("Shape:", orders.shape)

Saved to: ../data/processed/orders.parquet
Shape: (19213, 7)


## Exploratory Analysis

### Revenue Distribution

In [20]:
# ─────────────────────────────────────────────────────────────
# Revenue Distribution — Descriptive Statistics
# ─────────────────────────────────────────────────────────────

orders["total_revenue"].describe()

count    19213.000000
mean       459.688923
std        931.129071
min          0.840000
25%        159.380000
50%        303.540000
75%        484.440000
max      44051.600000
Name: total_revenue, dtype: float64

Revenue is **heavily right-skewed** — median order value (approx. 303 dollars) is well
below the mean (approx. 460 dollars), with a maximum exceeding 44,000 dollars. This long tail
is consistent with the presence of bulk purchasers and high-frequency buyers.

**Implication for modeling:** Raw revenue features may need log transformation
or careful normalization to prevent high-value outliers from dominating
model training.

In [21]:
# ─────────────────────────────────────────────────────────────
# Orders per Customer — Descriptive Statistics
# ─────────────────────────────────────────────────────────────

orders.groupby("Customer ID").size().describe()

count    4312.000000
mean        4.455705
std         8.170213
min         1.000000
25%         1.000000
50%         2.000000
75%         5.000000
max       205.000000
dtype: float64

Order frequency per customer is similarly skewed — the median customer
places only 2 orders, 75% of customers place 5 or fewer, and the most
active customer placed 205 orders across the dataset period.

---

## High-Level Observations

### Revenue Distribution

Customer revenue is **highly right-skewed**, with a small number of customers
accounting for a disproportionately large share of total revenue. Median order
revenue is substantially lower than the mean, indicating a long tail of
high-value orders consistent with bulk purchases or high-frequency buyers.

**Implications for modeling:** Revenue-based features may benefit from log
scaling. Model performance should be evaluated across customer segments to
ensure robustness across low- and high-value buyers.

---

### Orders per Customer

Order frequency is also **heavily skewed** — most customers place very few
orders while a small minority place dozens or hundreds. This long-tail
behavior is typical of retail businesses where a core group of highly engaged
users coexists with many sporadic or one-time buyers.

**Implications for modeling:** Frequency-based features such as rolling order
counts are likely to be strong predictors of short-term engagement. Customers
with limited order history represent a cold-start challenge that simple
heuristics cannot address, motivating the use of learned time-aware models.

---

### Overall Takeaway

The temporal structure and behavioral heterogeneity of this dataset provide
rich signal for churn modeling despite the relatively small number of raw
columns. These observations directly motivate:

- Rolling time-window features (7/30/90 days) to capture behavioral recency
  and frequency at multiple horizons
- Snapshot-based modeling rather than static customer aggregation
- Careful comparison against simple recency heuristics to demonstrate the
  value of a learned model